prow/pod-utils/gcs/upload_test.go

```diff
@@ -78,6 +94,15 @@ func TestUploadToGcs(t *testing.T) {
 			return nil
 		}
 
+		flaky := func(obj *storage.ObjectHandle) error {
```

Comment:
IMO the bigger issue with this unit test is that success, fail and flaky aren't part of the test case definition. I would refactor things to look more like this:

```golang

// change update func to set the name of the thing it uploaded

cases := []struct {
  name string
  uploads map[string]UploadFunc
  err bool
}{
  { name: "successful uploads work",
    uploads: map[string]UploadFunc{
      "hooray": func(o *s.OH) error { update("hooray"); return nil},
    },
  },
  {
    name: "failing uploads don't work", 
    ...
  }
}

for _, tc := range cases {
  err := Upload(tc.uploads)
  if err != nil { if !tc.err { t.errorf('boo") } }
  else if tc.err { // didn't raise an error }
  else {
    for key := range tc.uploads {
      // ensure update(key) was called by checking key is in the set that update updates (rather than count++)
}
```

But I don't think this PR needs to do that refactor, I'm fine with adding flaky alongside success

Reply:
I do not see why we need separate definitions of success and failure for each test case.

Reply:
I'm fine to rework this PR anyway you want. I am learning something new about how things work every iteration. Whatever you think is best

Reply:
The basic point here is that both of these are strong indications the test uses the wrong abstraction:
* every scenario requires additions outside the scenario (such as adding a new flaky function)
* every scenario provides the exact same coverage (all validate passing, failing, flaky functionality)

Assuming you care about validating the correct files were uploaded, each upload function needs to be unique so that it can note what was uploaded (aka `update()` needs to be `sets.Insert("unique-file")` not `counter++`). Thus one way or another each method therefore needs to track its name.

The current testCases structure is rather pointless, since the only thing you can control is the integer number of passing uploads, which is the wrong abstraction as there are at best only three classes:
* no passing uploads (negative passing is invalid)
* 1 passing upload
* multiple passing uploads (not really sure this is a meaningful difference from the 1 passing upload scenario)

Same for the other classes.

This encourages people to write meaningless values like `passingTargets: 5` which provides exactly the same coverage as the previous scenario that set 2, 1 or 1000.

You only need the passing function once (`passingTargets: N), or possibly twice (N = 1, N > 1), there's no reason not to put it in the test scenario. Putting it into the test scenario makes it easier to read, since the scenario setup, implementation and expectations all live right next to each other.

The scenarios should not be:
* "upload 2 passing, 3 failing, 1 flaky files correctly" 
* "upload 0 passing, 6 failing, 17 flaky files correctly"

But rather:
* "failing uploads report an error"
* "flaky uploads eventually succeed"
* "passing uploads complete"
* "complicated upload scenario works as expected"

This naturally suggests a structure like:
```golang
cases := []struct {
  name string
  fakeUploadFunc UploadFunc
  err bool
}
```

And if you want to handle the complex, multi file upload scenario:
```golang
cases := []struct {
  name string
  fakeUploadFuncs map[string]UploadFunc
  err bool
}
```

Reply:
this would be sort of an entire re-write of these tests, right? I can attempt that, but I know it will probably take some time and probably a lot of iterations with your review comments. should we get this current PR merged first, so we can have the functionality of the retries (to hopefully help
with flaky uploads in real tests), and I can work on a new PR to re-write things like @fejta
suggests.

or, if it's not really worth the effort of a re-write, you can suggest some other beginner-level
issue for me to work on.

Reply:
Correct, as the problems with this test method predate this PR, I'm more than happy to add the missing coverage using the same pattern, and use a separate PR to refactor the test method design (and awesome if that's something you're willing to do :hugs: )